fml/cpu_affinity.cc

```diff
@@ -0,0 +1,119 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/fml/cpu_affinity.h"
+
+#include <optional>
+#include <sstream>
+#include <string>
+
+namespace fml {
+
+CPUSpeedTracker::CPUSpeedTracker(std::vector<CpuIndexAndSpeed> data)
+    : cpu_speeds_(std::move(data)) {
+  std::optional<int32_t> max_speed = std::nullopt;
+  std::optional<int32_t> min_speed = std::nullopt;
+  for (const auto& data : cpu_speeds_) {
+    if (!max_speed.has_value() || data.speed > max_speed.value()) {
+      max_speed = data.speed;
+    }
+    if (!min_speed.has_value() || data.speed < min_speed.value()) {
+      min_speed = data.speed;
+    }
+  }
+  if (!max_speed.has_value() || !min_speed.has_value() ||
+      min_speed.value() == max_speed.value()) {
+    return;
+  }
+
+  for (const auto& data : cpu_speeds_) {
+    if (data.speed == max_speed.value()) {
+      performance_.push_back(data.index);
+    } else {
+      not_performance_.push_back(data.index);
+    }
+    if (data.speed == min_speed.value()) {
+      efficiency_.push_back(data.index);
+    }
+  }
+
+  valid_ = true;
+}
+
+bool CPUSpeedTracker::IsValid() const {
+  return valid_;
+}
+
+const std::vector<size_t>& CPUSpeedTracker::GetIndices(
+    CpuAffinity affinity) const {
+  switch (affinity) {
+    case CpuAffinity::kPerformance:
+      return performance_;
+    case CpuAffinity::kEfficiency:
+      return efficiency_;
+    case CpuAffinity::kNotPerformance:
+      return not_performance_;
+  }
+}
+
+// Get the size of the cpuinfo file by reading it until the end. This is
+// required because files under /proc do not always return a valid size
+// when using fseek(0, SEEK_END) + ftell(). Nor can they be mmap()-ed.
+std::optional<int32_t> ReadIntFromFile(const std::string& path) {
```

Comment:
This is scary looking but we're running lsan and asan over the tests. Unfortunately cant use the fml file APIs due to the note above.

Reply:
(also lsan and asan helped me fix several bugs here 😆 )

Reply:
Can `std::ifstream` read from `/proc` and `/sys`? If it can, I suspect that would simplify this code.

If it can't, I'll do a close review of this code =)

Reply:
Yeah, `std::` was/is banned from the Dart VM, but for new code in the Engine, I believe it is not banned =) cc @chinmaygarde in case he has different advice.

Reply:
ifstream appears to work!